readme.txt

GUI for MOAI
Derick Dong (derickdong@hotmail.com)

This is a fairly simple GUI system, but should be enough to handle the necessary tasks of most
developers. Its still in an early state, so expect bugs, quirks, and incomplete features.

There is currently support for the following widget types:

- Button - A basic push button
- Check Box - Can be set to checked or unchecked
- Edit Box - Allows entry of text
- Image - A static image
- Label - Simple static text
- Progress Bar - Displays current progress
- Radio Button - Can be selected or unselected; only one in the group may be selected at a time
- Scroll Bar - Scrolls through a list of items
- Slider - Selection of a numeric value by adjusting the slider
- Text Box - Displays multiple lines of text, with a scrollbar
- Widget List - Like a list box; supports multiple columns and multiple selections; different widget
	types can be associated with each column
- Window - A basic window; more of a container for other widgets than anything else


Some things to note:

- The GUI layer needs to be the topmost layer in order to receive input properly. Otherwise, layers
above will intercept it. The system creates its own layer, but will need to be added manually to
MOAIRenderMgr. The GUI layer can be retrieved by calling GUI:layer().

- The system uses a relative scale for the positioning and sizing of widgets, ranging from 0 to 100.
A widget's anchor point is at its top left. A position of (0, 0) would set the anchor to the top left
of the window, (100, 100) would be the lower right (and would normally take it out of the view area),
and (50, 50) would be the dead center. A size of (50, 50) makes the widget half the dimensions of the
screen in both width and height (ie. if the screen were (320, 480), the widget would be (160, 240)).


releasenotes.txt

0.2 - 04/28/12
Changes
- With the addition of MOAIRenderMgr, the system no longer adds its layer automatically to
  MOAI's rendering queue. This should solve the problem with scene managers
- More consistent function naming and parameters
- Added texture, font, and text style managers
- Using MOAIProp's attribute linking to maintain position information
- Removed the use of module(..., package.seeall)
- Using MOAITextStyles for text instead of just MOAIFont

Bugfixes
- Window widgets were not being created correctly by factory.lua
- Textbox - fixed a problem where the last "word" would not be displayed if it was only one character

0.201 - 05/08/12
Bugfixes
- Changed gui's _calcInputOver function to use MOAIPartition:propListForPoint instead of
  MOAIPartition:propForPoint. This will prevent windows that are hidden from accepting input


various.lua

--[[
	Demo for a basic GUI system

	This demonstrates creating various widgets programmatically.
]]

require "gui/support/class"

local gui = require "gui/gui"
local resources = require "gui/support/resources"
local filesystem = require "gui/support/filesystem"
local inputconstants = require "gui/support/inputconstants"
local layermgr = require "layermgr"

ButtonHandler = class()

-- Respond to button 1 being clicked
function _handleButton1Click(event, data)
	print("Button 1 Click")
end

-- Respond to button 2 being clicked
function ButtonHandler:handleButtonClick(event, data)
	print("Button 2 Click")
end

function ButtonHandler:init()

end

-- Turn off the texture loading logging
-- MOAILogMgr.setLogLevel(MOAILogMgr.LOG_NONE)

-- Setup your basic window
local width = 320
local height = 480

MOAISim.openWindow("Various", width, height)

-- Create the GUI, passing in the dimensions of the screen
local g = gui.GUI(width, height)

-- Resource paths - search through these for specified resources
g:addToResourcePath(filesystem.pathJoin("resources", "fonts"))
g:addToResourcePath(filesystem.pathJoin("resources", "gui"))
g:addToResourcePath(filesystem.pathJoin("resources", "media"))
g:addToResourcePath(filesystem.pathJoin("resources", "themes"))

-- Add the gui layer to the rendering stack, making sure its always high up on the
-- layer stack
layermgr.addLayer("gui", 99999, g:layer())

-- This sets up the theme to be used for widgets. If images aren't set
-- manually for a widget, the system looks to the theme for the images
-- to use.
g:setTheme("basetheme.lua")

-- The font used for text
g:setCurrTextStyle("default")

-- Show some simple text
local label = g:createLabel()
label:setPos(5, 3)
label:setDim(50, 5)
label:setText("Just some simple text")

-- A couple buttons
local button = g:createButton()
button:setPos(5, 10)
button:setDim(30, 10)
button:setText("button 1")

-- -- Register the button click event for this widget. In this case, we are registering to a
-- -- standard function, so the second parameter is 'nil'
button:registerEventHandler(button.EVENT_BUTTON_CLICK, nil, _handleButton1Click)

local buttonHandler = ButtonHandler()
local button = g:createButton()
button:setPos(45, 10)
button:setDim(30, 20)
button:setText("image button")
button:setNormalImage(resources.getPath("cathead.png"), 1, 1, 1, 1)
button:setHoverImage(resources.getPath("cathead.png"), 1, 0, 0, 1)
button:setPushedImage(resources.getPath("cathead.png"), 0, 1, 0, 1)
button:setDisabledImage(resources.getPath("cathead.png"), 0, 0, 1, 1)

-- Register the button click event for this widget. This time, we are registering to a
-- method of a class, so the second parameter is the class instance, and the function is
-- referenced by name (so the object can call it by table index)
button:registerEventHandler(button.EVENT_BUTTON_CLICK, buttonHandler, "handleButtonClick")

-- Static image
local image = g:createImage()
image:setPos(5, 23)
image:setDim(22, 15)
image:setText("Image")
image:setImage(resources.getPath("cathead.png"), 1, 1, 1, 1)
image:setTextAlignment(image.TEXT_ALIGN_CENTER, image.TEXT_ALIGN_BOTTOM)

-- -- Progress bar.
-- local progress = g:createProgressBar()
-- progress:setPos(45, 33)
-- progress:setDim(30, 5)
-- progress:setProgress(40)

-- A couple radio buttons
for i = 1, 3 do
	local radio = g:createRadioButton()
	radio:setPos((i - 1) * 25 + 5, 40)
	radio:setDim(20, 4)
	radio:setText("Radio")

	-- Assign all of these to the same group, so only one in this group can be selected at a time
	radio:setGroup(1)
end

-- Some check boxes
for i = 1, 3 do
	local check = g:createCheckBox()
	check:setPos((i - 1) * 25 + 5, 50)
	check:setDim(20, 4)
	check:setText("Check")
end

local label = g:createLabel()
label:setPos(5, 60)
label:setDim(30, 5)
label:setText("Slide Me")

-- A horizontal slider, with its current value displayed below the slider
local slider = g:createHorzSlider()
slider:setPos(35, 60)
slider:setDim(50, 5)
slider:setRange(1, 20)

local label = g:createLabel()
label:setPos(5, 70)
label:setDim(30, 5)
label:setText("Enter Text")

-- Allows entry of keyboard input
local edit = g:createEditBox()
edit:setPos(35, 70)
edit:setDim(50, 5)

-- Callback functions for input
-- When the user provides input, it must be sent into the GUI system so that it can
-- be handled properly.
function onKeyboardEvent(key, down)
	if (down == true) then
		g:injectKeyDown(key)
	else
		g:injectKeyUp(key)
	end
end

MOAIInputMgr.device.keyboard:setCallback ( onKeyboardEvent )

function onPointerEvent(x, y)
	g:injectMouseMove(x, y)
end

function onMouseLeftEvent(down)
	if (down) then
		g:injectMouseButtonDown(inputconstants.LEFT_MOUSE_BUTTON)
	else
		g:injectMouseButtonUp(inputconstants.LEFT_MOUSE_BUTTON)
	end
end

function onMouseMiddleEvent(down)
	if (down) then
		g:injectMouseButtonDown(inputconstants.MIDDLE_MOUSE_BUTTON)
	else
		g:injectMouseButtonUp(inputconstants.MIDDLE_MOUSE_BUTTON)
	end
end

function onMouseRightEvent(down)
	if (down) then
		g:injectMouseButtonDown(inputconstants.RIGHT_MOUSE_BUTTON)
	else
		g:injectMouseButtonUp(inputconstants.RIGHT_MOUSE_BUTTON)
	end
end

-- Register the callbacks for input
MOAIInputMgr.device.pointer:setCallback(onPointerEvent)
MOAIInputMgr.device.mouseLeft:setCallback(onMouseLeftEvent)
MOAIInputMgr.device.mouseMiddle:setCallback(onMouseMiddleEvent)
MOAIInputMgr.device.mouseRight:setCallback(onMouseRightEvent)




